homeassistant/components/device_tracker/mqtt.py

```diff
@@ -36,8 +36,11 @@ def async_setup_scanner(hass, config, async_see, discovery_info=None):
     @callback
     def async_tracker_message_received(topic, payload, qos):
         """Handle received MQTT message."""
-        hass.async_add_job(
-            async_see(dev_id=dev_id_lookup[topic], location_name=payload))
+        for subscription in dev_id_lookup:
+            if mqtt.match_topic(subscription, topic):
```

Comment:
Why not just create a separate callback for each subscription?

Reply:
Like this:

```python
    for dev_id, topic in devices.items():
        @callback
        def async_tracker_message_received(topic, payload, qos):
            """Handle received MQTT message."""
            hass.async_add_job(
                async_see(dev_id=dev_id, location_name=payload))

        yield from mqtt.async_subscribe(
            hass, topic, async_tracker_message_received, qos)
```

Reply:
Is it good style to pass the loop-local dev_id variable into a callback?
What value has the callback dev_id? The one from the loop or something unpredictable?

Reply:
That is indeed true, this wouldn't work
![image](https://user-images.githubusercontent.com/6833237/36000012-e8f4c378-0d21-11e8-9e3d-c5a383fedacd.png)

But using a callback creator should work:

```python
def make_callback(dev_id):
    @callback
    def async_tracker_message_received(...):
        # [...]
    return async_tracker_message_received

for dev_id, topic in devices.items():
    yield from mqtt.async_subscribe(hass, topic, make_callback(dev_id), qos)
```

I believe this would still be better than handling something yourself, that `mqtt` already handles and knows best how to handle.

Reply:
According to https://stackoverflow.com/a/3431699/3775041 this should also work:

```python
    for dev_id, topic in devices.items():
        @callback
        def async_tracker_message_received(topic, payload, qos, dev_id=dev_id):
            """Handle received MQTT message."""
            hass.async_add_job(
                async_see(dev_id=dev_id, location_name=payload))

        yield from mqtt.async_subscribe(
            hass, topic, async_tracker_message_received, qos)
```

Reply:
But you are right, the mqtt-outside handling of topics is not the best idea.

Reply:
BTW the old code also uses this strange callback-outside variable style with the dev_id_lookup dictionary.